server/chat-plugins/lottery.js

```diff
@@ -0,0 +1,248 @@
+'use strict';
+
+const LOTTERY_FILE = 'config/chat-plugins/lottery.json';
+
+/** @type {typeof import('../../lib/fs').FS} */
+const FS = require(/** @type {any} */('../../.lib-dist/fs')).FS;
+
+/** @type {{[roomid: string]: {maxWinners: number, name: string, markup: string, participants: {[ip: string]: string}}}} */
+let lotteries = {};
+try {
+	lotteries = require(`../../${LOTTERY_FILE}`);
+} catch (e) {
+	if (e.code !== 'MODULE_NOT_FOUND' && e.code !== 'ENOENT') throw e;
+}
+if (!lotteries || typeof lotteries !== 'object') lotteries = {};
+
+class Lottery {
+	/**
+     * @param {string} roomid
+     */
+	constructor(roomid) {
+		this.roomid = roomid;
+	}
+	/**
+	 * @param {number} maxWinners
+	 * @param {string} name
+	 * @param {string} markup
+	 */
+	create(maxWinners, name, markup) {
+		if (!lotteries[this.roomid]) {
+			lotteries[this.roomid] = {maxWinners, name, markup, participants: {}};
+			this.write();
+		}
+	}
+	write() {
+		FS(LOTTERY_FILE).writeUpdate(() => JSON.stringify(lotteries));
+	}
+	destroy() {
+		delete lotteries[this.roomid];
+		this.write();
+	}
+	/**
+     * @param {User} user
+     */
+	addUser(user) {
+		if (!lotteries[this.roomid].participants[user.latestIp]) {
+			lotteries[this.roomid].participants[user.latestIp] = user.name;
+			this.write();
+			return true;
+		}
+		return false;
+	}
+	/**
+     * @param {User} user
+     */
+	removeUser(user) {
+		if (lotteries[this.roomid].participants[user.latestIp]) {
+			delete lotteries[this.roomid].participants[user.latestIp];
+			this.write();
+			return true;
+		}
+		return false;
+	}
+	getWinner() {
+		const winners = [];
```

Comment:
I know that you _can_ push to arrays that are delcared as a ``const``, but I think I was told a while ago that we should still declare them using ``let`` because their value is subject to change

Reply:
I feel like I've seen this style of coding before though. I also can't find it in ``CONTRIBUTING.md``, do you have a link?

Reply:
Unfortunately, I do not. I think it was @Zarel who told me this, though.

Reply:
I never put it in CONTRIBUTING, but I did _write_ in that style for a while.

I've given up on that style of coding, because it conflicts with `prefer-const`. And I've never actually inflicted it on other people in code review.

Reply:
If I told you, it's probably because someone asked what I was doing in my own code I wrote myself, and I was explaining that. I have no memory of telling anyone else to do it in that style.

Reply:
I found the discussion for reference, it was on https://github.com/Zarel/Pokemon-Showdown/commit/1a513ed585f755200d2777c2e45439672ce16223